### PR TITLE
chore(core): Migrate `LogSchema` `timestamp` to new lookup code

### PR DESCRIPTION
**Pull request by fuchsnj**
fixes https://github.com/vectordotdev/vector/issues/16834

The `LogSchema` timestamp key was previously a `String`, which makes it very easy to mis-use it. It's not obvious that it can be optional (by using an empty string, or technically any invalid path). This has been migrated to a typed path which should help prevent these types of problems from happening again.

### CODE REVIEW

File: lib/codecs/src/decoding/format/gelf.rs [hunk 1]
```diff
@@ -98,19 +98,21 @@ impl GelfDeserializer {
...
-            )
-            .expect("invalid timestamp");
-            log.insert(
-                log_schema().timestamp_key(),
-                DateTime::<Utc>::from_utc(naive, Utc),
-            );
-        // per GELF spec- add timestamp if not provided
-        } else {
-            log.insert(log_schema().timestamp_key(), Utc::now());
+        if let Some(timestamp_key) = log_schema().timestamp_key() {
```
Comment (spencergilbert) [inline lib/codecs/src/decoding/format/gelf.rs:101]: This conditional gets dropped again when/if we remove the global log schema?

File: src/sinks/aws_kinesis/streams/integration_tests.rs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #![cfg(feature = "aws-kinesis-streams-integration-tests")]
 #![cfg(test)]
+#![allow(warnings)]
```
Comment (spencergilbert) [inline src/sinks/aws_kinesis/streams/integration_tests.rs:3]:
```suggestion
```

File: src/sinks/aws_kinesis/streams/integration_tests.rs [hunk 3]
```diff
@@ -23,62 +25,66 @@ fn kinesis_address() -> String {
...
+    let sink = config.build(cx).await.unwrap().0;
+
+    let timestamp = chrono::Utc::now().timestamp_millis();
+
+    let (mut input_lines, events) =
+        random_updated_events_with_stream(100, 11, None, |(i, mut event)| {
+            event.insert(partition_key.as_str(), "abcd");
+            event
+        });
+    // = random_lines_with_stream(100, 11, None);
```
Comment (spencergilbert) [inline src/sinks/aws_kinesis/streams/integration_tests.rs:67]:
```suggestion
```

File: src/sinks/aws_kinesis/streams/integration_tests.rs [hunk 4]
```diff
@@ -23,62 +25,66 @@ fn kinesis_address() -> String {
...
+        .map(|e| {
+            assert_eq!(Some(partition_key.as_str()), e.partition_key());
+            e
+        })
+        .map(|e| String::from_utf8(e.data.unwrap().into_inner()).unwrap())
+        .collect::<Vec<_>>();
+
+    // input_lines.sort();
+    // output_lines.sort();
+    // assert_eq!(output_lines, input_lines)
```
Comment (spencergilbert) [inline src/sinks/aws_kinesis/streams/integration_tests.rs:86]:
```suggestion]
```

**Review by spencergilbert — CHANGES_REQUESTED** [review]
Caught a couple of things 😄

File: scripts/integration/aws/test.yaml [hunk 5]
```diff
@@ -1,7 +1,7 @@
 features:
 - aws-integration-tests
 
-test_filter: ::aws_
+test_filter: ::aws_kinesis::streams
```
[the remark this answers is not in the document]
Comment (fuchsnj) [reply scripts/integration/aws/test.yaml:4]: oops, yea the integration changes shouldn't have been included here

**Comment by spencergilbert** [comment]
Did you intentionally leave that kinesis integration test commented out again?

**Comment by fuchsnj** [comment]
> Did you intentionally leave that kinesis integration test commented out again?

Yes, that was just testing I was doing locally. It shouldn't have been included in this PR since it's completely unrelated.

**Review by spencergilbert — APPROVED** [review]

**Review by StephenWakely — APPROVED** [review]
Looks good to me.

**Pull request merged**